### PR TITLE
Basic CTL shared_ptr implementation

### PR DESCRIPTION
**Pull request by mrdomino**

### CODE REVIEW

**Comment by mrdomino** [comment]
Supersedes #1229.

File: ctl/shared_ptr.h [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+};
+
+static inline __attribute__((always_inline)) void
+incref(size_t* r) noexcept
+{
+#ifdef NDEBUG
+    __atomic_fetch_add(r, 1, __ATOMIC_RELAXED);
+#else
+    size_t refs = __atomic_fetch_add(r, 1, __ATOMIC_RELAXED);
+    if (refs > ((size_t)-1) >> 1)
```
Comment (jart) [inline ctl/shared_ptr.h:47]: Is there any particular reason why you wouldn't just use `long` and then check if it's negative?

**Review by jart — APPROVED** [review]
Looks great. We know this algorithm is correct, but if you want to make extra sure, you could add a `pthread_create` torture test. I have lots of them in the codebase.

File: ctl/shared_ptr.h (hunk 1 above)
Comment (mrdomino) [reply ctl/shared_ptr.h:47]: You’d need `ckd_add` or something in that case - I believe signed integer overflow is UB

Comment (mrdomino) [reply ctl/shared_ptr.h:47]: Oh wait, no, I see what you mean. Yeah, that could work.

**Pull request merged**